prov/usnic/src/usdf_mem.c

```diff
@@ -80,13 +80,14 @@ static struct fi_ops usdf_mr_ops = {
 };
 
 int
-usdf_reg_mr(struct fid_domain *domain, const void *buf, size_t len,
+usdf_reg_mr(struct fid *fid, const void *buf, size_t len,
 	   uint64_t access, uint64_t offset, uint64_t requested_key,
 	   uint64_t flags, struct fid_mr **mr_o, void *context)
 {
 	struct usdf_mr *mr;
 	struct usdf_domain *udp;
 	int ret;
+	struct fid_domain *domain = container_of(fid, struct fid_domain, fid);
```

Comment:
Shouldn't this at least be rewritten this way?  (same for the similar changes in the other providers)

```c
struct fid_domain *domain;

assert(fid->fclass == FI_CLASS_DOMAIN);
domain = container_of(fid, struct fid_domain, fid);
```

Or:

```c
struct fid_domain *domain;

if (fid->fclass != FI_CLASS_DOMAIN) {
        USDF_DEBUG("memory registration only supported for struct fid_domain\n");
        return -FI_EINVAL;
}
domain = container_of(fid, struct fid_domain, fid);
```

If future client code starts passing non-domain objects and someone runs that client code with an older libfabric (from after this change but before future libfabric starts supporting non-domain objects for registration), bad things would start to happen with no explanation without an error checking addition.